### ISSUE TITLE
Fix scripts/check_style.sh to work with latest Github API

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#10190]
From https://app.circleci.com/pipelines/github/ethereum/solidity/9790/workflows/c58049a5-b961-4e6b-8ceb-4d47fa3a7168/jobs/487386:
```
#!/bin/bash -eo pipefail
./scripts/check_style.sh
Error: Trailing whitespace found:
test/libsolidity/semanticTests/underscore/as_function.sol:5:    
test/libsolidity/semanticTests/underscore/in_modifier.sol:11:    
test/libsolidity/syntaxTests/underscore/as_function.sol:4:    
test/libsolidity/syntaxTests/underscore/in_modifier.sol:11:    
Error: Trailing whitespace found: test/libsolidity/semanticTests/underscore/as_function.sol:5: test/libsolidity/semanticTests/underscore/in_modifier.sol:11: test/libsolidity/syntaxTests/underscore/as_function.sol:4: test/libsolidity/syntaxTests/underscore/in_modifier.sol:11:
posting error message to github
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}

Exited with code exit status 1
CircleCI received exit code 1
```

**Comment by axic** [comment]
I guess this was solved as we haven't seen it lately?

**Comment by cameel** [comment]
Still seems to be happening. Just looking at [the most recent PR with failed `chk_coding_style` job](https://app.circleci.com/pipelines/github/ethereum/solidity/12545/workflows/7bc69c4c-cd03-4ef7-86d8-567892a18c63/jobs/588484) (#10777):
```
Error: Trailing whitespace found:
test/libsolidity/smtCheckerTests/bmc_coverage/unchecked_functional_call_with_unchecked_block.sol:7:		}    
test/libsolidity/smtCheckerTests/unchecked/unchecked_functional_call_with_unchecked_block.sol:7:		}    
Error: Trailing whitespace found: test/libsolidity/smtCheckerTests/bmc_coverage/unchecked_functional_call_with_unchecked_block.sol:7: } test/libsolidity/smtCheckerTests/unchecked/unchecked_functional_call_with_unchecked_block.sol:7: }
posting error message to github
{
  "message": "Requires authentication",
  "documentation_url": "https://docs.github.com/rest/reference/issues#create-an-issue-comment"
}
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/reference/pulls#create-a-review-comment-for-a-pull-request"
}
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/reference/pulls#create-a-review-comment-for-a-pull-request"
}

Exited with code exit status 1

CircleCI received exit code 1
```

I also do not see these messages posted as comments in the PR but I've seen comments from the bot appear in other PRs so it's broken in some way but not completely.

**Comment by blishko** [comment]
Yeah, I noticed the difference in my PRs and e.g. Leo's PRs where he gets nice comments from bot and I don't.
I assumed it's because I don't have the same level of permissions as he does.
Or is it something else?

**Comment by cameel** [comment]
Hmm... I don't think that's the issue, but I'm not 100% sure. [`report_errors.sh`](https://github.com/ethereum/solidity/blob/develop/scripts/report_errors.sh) posts the comments as `stackenbotten` with `$GITHUB_ACCESS_TOKEN` as password. The token comes from job environment so I think it does not depend on the account of the PR author but rather comes from CircleCI settings.

I'll take a closer look at this later today. I suspect that it's just some simple error in the script that's triggered only with some specific kind of errors - maybe something unexpected contains whitespace - but we'll see.

**Comment by ekpyron** [comment]
I wouldn't be surprised if permission levels play a role here - such access tokens are at times used for more serious stuff than posting comments, so being able to just create a PR from a fork that can just change the script that uses the token to do whatever can be dangerous at times, can't it :-)?

**Comment by cameel** [comment]
Looks like it's not about permissions. Or not only about permissions. I just got that error in a PR started by @axic: https://app.circleci.com/pipelines/github/ethereum/solidity/12578/workflows/f9de0575-56ad-4bbc-b180-343ed4bc0182/jobs/589674

And this time there's actually a bit more concrete info:
```json
{
  "message": "Validation Failed",
  "errors": [
    {
      "resource": "PullRequestReviewComment",
      "code": "invalid",
      "field": "path"
    }
  ],
  "documentation_url": "https://docs.github.com/rest/reference/pulls#create-a-review-comment-for-a-pull-request"
}
```

Looks like there's something wrong with `\"path\": \"$ERROR_PATH\"` in the script, which comes from `ERROR_PATH=$(echo $line | grep -oE ".*\.cpp")`. My errors are all about `*.sol` files - maybe that's the issue?

There also seems to be some problem with the file being checked multiple times in parallel, which messes out the output:
```
test/libsolidity/syntaxTests/license/license_whitespace_trailing.sol:1:// SPDX-License-Identifier: GPL-3.0   
 test/libsolidity/syntaxTests/license/license_whitespace_trailing.sol:1:// SPDX-License-Identifier: GPL-3.0est suite adds the automatic preamble.te adds the automatic preamble.
```

**Comment by cameel** [comment]
> I wouldn't be surprised if permission levels play a role here - such access tokens are at times used for more serious stuff than posting comments, so being able to just create a PR from a fork that can just change the script that uses the token to do whatever can be dangerous at times, can't it :-)?

That's a good point. But then I wonder which token it would use? If it's the one from PR creator then what if someone pushes extra commits to someone else's branch? Anyway, I'll check the script first because I think the problem might be much simpler than that. I really don't like how this script is organized (it requires you to source it and then runs `trap` to get executed just when the parent script exits) and on top of that I don't think it handles all errors and corner cases properly.

**Comment by cameel** [comment]
I have a fix in #10792. There were two problems:
1. For some reason the script runs twice in CI. I don't know exactly why, but it's related to its weird flow. Before my changes `check_style.sh` would source it and it would set `trap EXIT` to get called when `check_style.sh` exits. Changing it to a normal style where `check_style.sh` just calls it in a normal way solved that problem. It now runs only once.
2. Script was hard-coded to store the style errors in `/tmp/error.log` and was not clearing it between runs. The first line of this file contains the text `Coding style error:` and subsequent lines are the style errors. The script needs to remove that first line - otherwise it would get empty values for file path and line number. The problem was that, because of multiple simultaneous runs, there were multiple such lines and this caused the script to send invalid JSON with missing values to github. It does not run with `set -e` though so this did not prevent it from continuing and posting the actual style errors.

**Comment by cameel** [comment]
And here's another PR that cleans up these scripts a bit: #10797.

**Comment by cameel** [comment]
I just noticed one more case where the script fails: it's during the initial CI run that starts immediately when a branch is pushed, but before a PR is even created:

```
Coding style error:
libsolidity/interface/Natspec.cpp:59:	if(!notice.empty())
posting error message to github
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}
```

That one can't really be fixed, only hidden. We could detect the `Not Found` message and print something more informative than a JSON dump but I think it's just a minor annoyance so I'm leaving it as is unless someone else thinks this really needs be changed.
